<a href="https://colab.research.google.com/github/ThomasDarrieumerlou/Applied_Predictive_Cours/blob/main/NLP_KNN_BERT_EMBEEDINGS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# BERT Embeeding Natural Language Processing </a>

## K Nearest Neighbors Model for a Classification Problem: Classify Product Reviews as Positive or Negative

In this notebook, we use the K Nearest Neighbors method to build a classifier to predict the __isPositive__ field of our review dataset (that is very similar to the final project dataset).


1. <a href="#1">Reading the dataset</a>
2. <a href="#2">Exploratory data analysis</a>
3. <a href="#3">Text Processing: Stop words removal and stemming</a>
4. <a href="#4">Train - Validation Split</a>
5. <a href="#5">Data processing with Pipeline</a>
6. <a href="#6">Train the classifier</a>
7. <a href="#7">Test the classifier</a> Find more details on the KNN Classifier here: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html
8. <a href="#8">Ideas for improvement</a>

Overall dataset schema:
* __reviewText:__ Text of the review
* __summary:__ Summary of the review
* __verified:__ Whether the purchase was verified (True or False)
* __time:__ UNIX timestamp for the review
* __log_votes:__ Logarithm-adjusted votes log(1+votes). *This field is a processed version of the votes field. People can click on the "helpful" button when they find a customer review helpful. This increases the vote by 1. __log_votes__ is calculated like this log(1+votes). This formulation helps us get a smaller range for votes.*
* __isPositive:__ Whether the review is positive or negative (1 or 0)


## 1. <a name="1">Reading the dataset</a>
(<a href="#0">Go to top</a>)

We will use the __pandas__ library to read our dataset.

In [ ]:
import pandas as pd

#df = pd.read_csv('https://raw.githubusercontent.com/aws-samples/aws-machine-learning-university-accelerated-nlp/master/data/examples/AMAZON-REVIEW-DATA-CLASSIFICATION.csv')

#print('The shape of the dataset is:', df.shape)

In [ ]:
# IMDB Dataset
df = pd.read_csv('https://raw.githubusercontent.com/aws-samples/aws-machine-learning-university-accelerated-nlp/master/data/final_project/imdb_train.csv', header=0)

train_df = pd.read_csv('https://raw.githubusercontent.com/aws-samples/aws-machine-learning-university-accelerated-nlp/master/data/final_project/imdb_train.csv', header=0)
train_df.head()

test_df = pd.read_csv('https://raw.githubusercontent.com/aws-samples/aws-machine-learning-university-accelerated-nlp/master/data/final_project/imdb_test.csv', header=0)
test_df.head()


text  label
0  What I hoped for (or even expected) was the we...      0
1  Garden State must rate amongst the most contri...      0
2  There is a lot wrong with this film. I will no...      1
3  To qualify my use of "realistic" in the summar...      1
4  Dirty War is absolutely one of the best politi...      1

Let's look at the first 10 rows of the dataset.

In [ ]:
df.head(10)

reviewText  \
0  PURCHASED FOR YOUNGSTER WHO\nINHERITED MY "TOO...   
1                              unable to open or use   
2   Waste of money!!! It wouldn't load to my system.   
3  I attempted to install this OS on two differen...   
4  I've spent 14 fruitless hours over the past tw...   
5  I purchased the home and business because I wa...   
6  The download doesn't take long at all. And it'...   
7  This program is positively wonderful for word ...   
8   Fantastic protection!!  Great customer support!!   
9  Obviously Win 7 now the last great operating s...   

                                             summary  verified        time  \
0                                IDEAL FOR BEGINNER!      True  1361836800   
1                                          Two Stars      True  1452643200   
2                                       Dont buy it!      True  1433289600   
3  I attempted to install this OS on two differen...      True  1518912000   
4                                   Do NOT Download.      True  1441929600   
5        Quicken home and business  not for amatures      True  1335312000   
6                                             Great!      True  1377993600   
7                             Terrific for practice.     False  1158364800   
8                                         Five Stars      True  1478476800   
9                                         Five Stars      True  1471478400   

   log_votes  isPositive  
0   0.000000         1.0  
1   0.000000         0.0  
2   0.000000         0.0  
3   0.000000         0.0  
4   1.098612         0.0  
5   0.000000         0.0  
6   0.000000         1.0  
7   2.397895         1.0  
8   0.000000         1.0  
9   0.000000         1.0

## 2. <a name="2">Exploratory data analysis</a>
(<a href="#0">Go to top</a>)

Let's look at the distribution of __isPositive__ field.

In [ ]:
df["isPositive"].value_counts()

1.0    43692
0.0    26308
Name: isPositive, dtype: int64

We can check the number of missing values for each columm below.

In [ ]:
print(df.isna().sum())

text     0
label    0
dtype: int64


We have missing values in our text fields.

## 3. <a name="3">Text Processing: Stop words removal and stemming</a>
(<a href="#0">Go to top</a>)

In [ ]:
df=df.dropna()
print(df.isna().sum())

text     0
label    0
dtype: int64


In [ ]:
!pip install transformers torch scikit-learn


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.7/7.7 MB 28.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.0/302.0 kB 25.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 39.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 39.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 295.0/295.0 kB 27.7 MB/s eta 0:00:00


## 4. <a name="4">Train - Validation Split</a>
(<a href="#0">Go to top</a>)

Let's split our dataset into training (90%) and validation (10%).

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Split the dataset into train and test sets
#train_data, test_data, train_labels, test_labels = train_test_split(df["reviewText"], df['isPositive'], test_size=0.1, random_state=42)

train_data, test_data, train_labels, test_labels = train_test_split(df["text"], df['label'], test_size=0.1, random_state=42)

## Use BERT for text embeddings:
(<a href="#0">Go to top</a>)

You can use the Hugging Face Transformers library to load a pre-trained BERT model and tokenize your text dat

In [ ]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(device)

cuda


In [ ]:
#from transformers import BertTokenizer, BertModel
#from transformers import RobertaTokenizer, RobertaModel
from transformers import AlbertTokenizer, AlbertModel

import torch
from tqdm import tqdm


# Move the model to GPU if available
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
#model.to(device)
# BERT-based Classifier
# Load BERT tokenizer and model, move to GPU

#tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
#bert_model = BertModel.from_pretrained("bert-base-uncased").to(device)

tokenizer = RobertaTokenizer.from_pretrained('roberta-base')
bert_model = RobertaModel.from_pretrained('roberta-base').to(device)

tokenizer = AlbertTokenizer.from_pretrained("albert-base-v2")
bert_model = AlbertModel.from_pretrained("albert-base-v2").to(device)




# Tokenize and encode the text data, move to GPU
max_length = 128
X_train_tokens = tokenizer(list(train_data), truncation=True, padding=True, max_length=max_length, return_tensors="pt", add_special_tokens=True).to(device)
X_test_tokens = tokenizer(list(test_data), truncation=True, padding=True, max_length=max_length, return_tensors="pt", add_special_tokens=True).to(device)

# Calculate BERT embeddings for the text data
def get_bert_embeddings(tokens):
    embeddings = []
    for i in tqdm(range(len(tokens['input_ids']))):
        with torch.no_grad():
            #output = model(**X_train_tokens)
            output = bert_model(input_ids=tokens['input_ids'][i].unsqueeze(0), attention_mask=tokens['attention_mask'][i].unsqueeze(0))
        embeddings.append(output[0].squeeze().mean(dim=0).cpu().numpy())
    return embeddings

X_train_bert_embeddings = get_bert_embeddings(X_train_tokens)
X_test_bert_embeddings = get_bert_embeddings(X_test_tokens)

Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
100%|██████████| 2500/2500 [00:30<00:00, 82.54it/s]


## 6. <a name="6">Train the classifier</a>
(<a href="#0">Go to top</a>)

We train our classifier with __.fit()__ on our training dataset.
Train a KNN model:
Now that you have BERT embeddings for your text data, you can train a KNN model using scikit-learn.

This code demonstrates how to use BERT for text embeddings and then train a KNN model for sentiment analysis of AWS product reviews. Make sure to replace 'your_dataset.csv' with the actual path to your dataset file and adjust other parameters as needed. You may also fine-tune the model and preprocessing steps to improve performance.

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Initialize and train the KNN classifier
knn = KNeighborsClassifier(n_neighbors=10)
knn.fit(X_train_bert_embeddings, train_labels)

NB = GaussianNB()
NB.fit(X_train_bert_embeddings, train_labels)

# Train a classifier on BERT embeddings (you can use any classifier of your choice)
# Here, we'll use Logistic Regression as an example
lr = LogisticRegression(max_iter=1000)
lr.fit(X_train_bert_embeddings, train_labels)

#rf=RandomForestClassifier()
#rf.fit(train_embeddings, train_labels)
#xgb=GradientBoostingClassifier()
#xgb.fit(train_embeddings, train_labels)



LogisticRegression(max_iter=1000)

##Test the classifier
(Go to top)

To evaluate the KNN model's performance on sentiment classification, you can generate a classification report and a confusion matrix. Here's how you can do it using scikit-learn:
|--|--|--|

In [ ]:
from sklearn.metrics import classification_report, confusion_matrix

# Make predictions on the test data
predictions = knn.predict(X_test_bert_embeddings)

# Calculate accuracy
accuracy = accuracy_score(test_labels, predictions)
print(f'KNN Accuracy: {accuracy * 100:.2f}%')
#predictions = rf.predict(test_embeddings)
predictions = NB.predict(X_test_bert_embeddings)
accuracy = accuracy_score(test_labels, predictions)
print(f'Naive Bais Accuracy: {accuracy * 100:.2f}%')

#predictions = rf.predict(test_embeddings)
predictions = lr.predict(X_test_bert_embeddings)
accuracy = accuracy_score(test_labels, predictions)
print(f'Logistic Regression Accuracy: {accuracy * 100:.2f}%')

KNN Accuracy: 78.08%
Naive Bais Accuracy: 78.00%
Logistic Regression Accuracy: 85.80%


In [ ]:
from sklearn.metrics import classification_report, confusion_matrix

# Make predictions on the test data
predictions = knn.predict(X_test_bert_embeddings)
# Create a classification report
class_report = classification_report(test_labels, predictions, target_names=['negative', 'positive'])

# Create a confusion matrix
conf_matrix = confusion_matrix(test_labels, predictions)

# Print the classification report and confusion matrix
print("Classification Report:")
print(class_report)

print("\nConfusion Matrix:")
print(conf_matrix)

#predictions = rf.predict(test_embeddings)
predictions = NB.predict(X_test_bert_embeddings)
# Create a classification report
class_report = classification_report(test_labels, predictions, target_names=['negative', 'positive'])

# Create a confusion matrix
conf_matrix = confusion_matrix(test_labels, predictions)

# Print the classification report and confusion matrix
print("Classification Report:")
print(class_report)

print("\nConfusion Matrix:")
print(conf_matrix)

#predictions = rf.predict(test_embeddings)
predictions = lr.predict(X_test_bert_embeddings)
# Create a classification report
class_report = classification_report(test_labels, predictions, target_names=['negative', 'positive'])

# Create a confusion matrix
conf_matrix = confusion_matrix(test_labels, predictions)

# Print the classification report and confusion matrix
print("Classification Report:")
print(class_report)

print("\nConfusion Matrix:")
print(conf_matrix)




Classification Report:
              precision    recall  f1-score   support

    negative       0.72      0.89      0.80      1207
    positive       0.87      0.68      0.76      1293

    accuracy                           0.78      2500
   macro avg       0.79      0.78      0.78      2500
weighted avg       0.80      0.78      0.78      2500


Confusion Matrix:
[[1070  137]
 [ 411  882]]
Classification Report:
              precision    recall  f1-score   support

    negative       0.77      0.77      0.77      1207
    positive       0.79      0.79      0.79      1293

    accuracy                           0.78      2500
   macro avg       0.78      0.78      0.78      2500
weighted avg       0.78      0.78      0.78      2500


Confusion Matrix:
[[ 932  275]
 [ 275 1018]]
Classification Report:
              precision    recall  f1-score   support

    negative       0.85      0.85      0.85      1207
    positive       0.86      0.86      0.86      1293

    accuracy         

In [ ]:
from transformers import pipeline

classifier = pipeline("feature-extraction")
#cl=classifier("I've been waiting for a HuggingFace course  my whole life.")
cl=classifier("Lyon is the best city in France")

cl

No model was supplied, defaulted to distilbert-base-cased and revision 935ac13 (https://huggingface.co/distilbert-base-cased).
Using a pipeline without specifying a model name and revision in production is not recommended.


[[[0.3689537048339844,
   -0.041106365621089935,
   -0.24234828352928162,
   -0.20132727921009064,
   -0.19458365440368652,
   0.10969351977109909,
   0.24939468502998352,
   -0.09835966676473618,
   0.025061793625354767,
   -1.1011372804641724,
   -0.30452483892440796,
   0.07605492323637009,
   -0.04040786251425743,
   -0.013498580083251,
   -0.2441929131746292,
   0.13367751240730286,
   0.24318407475948334,
   0.2112209051847458,
   0.1023818701505661,
   -0.21913041174411774,
   0.0558667853474617,
   -0.22697670757770538,
   0.4339923858642578,
   -0.19047322869300842,
   0.15832141041755676,
   -0.207810640335083,
   0.0910896435379982,
   0.29036015272140503,
   -0.08799152076244354,
   0.3524472713470459,
   -0.08077921718358994,
   0.3583160936832428,
   -0.05953880399465561,
   0.18912719190120697,
   -0.21855853497982025,
   0.28817349672317505,
   0.09951124340295792,
   -0.502049446105957,
   -0.14428338408470154,
   -0.10557772219181061,
   -0.5113519430160522,
   0.0719

------------------------------------------------------------------------
=> Result obtained with BOW

Testing environment :
- Use with KNN TfidfVectorizer set at 50 feature

Result Table

| KNN Value | Accuracy use_idf TRUE | Accuracy use_idf FALSE |
| --------- | ------------- | -------------- |
| K = 3     | 0.639         | 0.63144        |
| k = 5     | 0.65116       | 0.64528        |
| k = 7     | 0.66192       | 0.6532         |
| k = 9     | 0.66768       | 0.65736        |
| k = 15    | 0.6762        | 0.66832        |

------------------------------------------------------------------------
=> Result obtained with BERT (AMAZON-REVIEW-DATA-CLASSIFICATION.csv)

Testing environment :
- Use with KNN = 10

Result Table

|Accuracy|
|--------|
| 0.84 |
| 0.66 |
| 0.88 |

Is better than BoW

=> Result obtained with Roberta (AMAZON-REVIEW-DATA-CLASSIFICATION.csv)

Testing environment :
- Use with KNN = 10

Result Table

|Accuracy|
|--------|
| 0.82 |
| 0.59 |
| 0.89 |


=> Result obtained with BERT (imdb_train.csv)

Testing environment :
- Use with KNN = 10

Result Table

|Accuracy|
|--------|
| 0.77 |
| 0.74 |
| 0.83 |

Is better than BoW

=> Result obtained with ROBERTA (imdb_train.csv)

Testing environment :
- Use with KNN = 10

Result Table

|Accuracy|
|--------|
| 0.78 |
| 0.78 |
| 0.86 |

Is better than BoW